### PR TITLE
builtins: add time.clock, time.date, and time.weekday

### PR DESCRIPTION
**Pull request by DFrenkel**
### What code changed, and why?
The following builtins have been implemented:
* [time.clock](https://www.openpolicyagent.org/docs/policy-reference/builtins/time#builtin-time-timeclock)
* [time.date](https://www.openpolicyagent.org/docs/policy-reference/builtins/time#builtin-time-timedate)
* [time.weekday](https://www.openpolicyagent.org/docs/policy-reference/builtins/time#builtin-time-timeweekday)

### Definition of done
Compliance tests pass:
```
✅ time/test-time-0951.json: time/date -> eval
✅ time/test-time-0952.json: time/date with LA tz -> eval
✅ time/test-time-0953.json: time/date with empty tz -> eval
✅ time/test-time-0954.json: time/date leap day -> eval
✅ time/test-time-0955.json: time/date too big -> eval
✅ time/test-time-0956.json: time/clock -> eval
✅ time/test-time-0957.json: time/clock with NY tz -> eval
✅ time/test-time-0958.json: time/clock leap day -> eval
✅ time/test-time-0959.json: time/clock too big -> eval
✅ time/test-time-0960.json: time/weekday -> eval
✅ time/test-time-0961.json: time/weekday -> eval
✅ time/test-time-0962.json: time/weekday -> eval
✅ time/test-time-0963.json: time/weekday -> eval
✅ time/test-time-0964.json: time/weekday -> eval
✅ time/test-time-0965.json: time/weekday -> eval
✅ time/test-time-0966.json: time/weekday -> eval
✅ time/test-time-0967.json: time/weekday too big -> eval
```

### How to test
```
make all
OPA_COMPLIANCE_TESTS="time" make test-compliance | grep -E "weekday|clock|date"
```

### Related Resources
Related to #116

### CODE REVIEW

File: Sources/Rego/Builtins/Time.swift [hunk 1]
```diff
@@ -71,6 +74,96 @@ extension BuiltinFuncs {
...
+        let (date, cal) = try dateWithCalendar(args[0])
+        return .string(weekdayNames[cal.component(.weekday, from: date) - 1])
+    }
+
+    // MARK: - private utilities
+    /// Parses `nanoseconds` or `[nanoseconds, timezone]` argument and returns the corresponding `Date` and
+    /// a UTC-based Gregorian `Calendar` configured with the specified timezone.
+    private static func dateWithCalendar(_ arg: AST.RegoValue) throws -> (Date, Calendar) {
+        let (ns, timeZone) = try parseNanosWithTimezone(arg)
+        let date = Date(timeIntervalSince1970: TimeInterval(ns / 1_000_000_000))
```
Comment (philipaconrad) [inline Sources/Rego/Builtins/Time.swift]: [comment]: I'd wondered why the ns timestamp was truncated so severely here. It turns out [`TimeInterval`](https://developer.apple.com/documentation/foundation/timeinterval) expects a timestamp in *seconds*, so the truncation makes sense in context. For date/calendar operations, I *think* that should be fine for our purposes.

I'd be curious to see an edge-case test for this though-- I'm not sure what the exactly correct behavior should be, for us or OPA. 🤔 If a timestamp is technically 9/10 of a second into the new year, we'd want to handle that case, right?

File: Tests/RegoTests/BuiltinTests/TimeTests.swift [hunk 2]
```diff
@@ -98,9 +406,24 @@ extension BuiltinTests.TimeTests {
...
+                argName: "x",
+                allowedArgTypes: ["array", "number[integer]"],
+                wantArgs: "any<number[integer], array<number[integer], string>>",
+                generateNumberOfArgsTest: true, numberAsInteger: true),
+            BuiltinTests.generateFailureTests(
+                builtinName: "time.date", sampleArgs: [[1_585_852_421_593_912_000, "UTC"]], argIndex: 0,
+                argName: "x",
+                allowedArgTypes: ["array", "number[integer]"],
+                wantArgs: "any<number[integer], array<number[integer], string>>",
+                generateNumberOfArgsTest: true, numberAsInteger: true),
```
Comment (philipaconrad) [inline Tests/RegoTests/BuiltinTests/TimeTests.swift:462]: [question]: Could we add the `generateFailureTests` for `time.weekday` as well?

**Review by philipaconrad — COMMENTED** [review]
Hello @DFrenkel! Overall, this PR looks great!

I've got a few comments and questions below:

File: Sources/Rego/Builtins/Time.swift (hunk 1 above)
Comment (DFrenkel) [reply Sources/Rego/Builtins/Time.swift]: If we get sub-second value of the timestamp, truncating it to a whole second should *not* change either the [year, month, date] or [hour, minute, second] value of it, and the weekday value, should it? I added tests for sure, thanks for pointing this out!
Also: turns out there *was* a bug in this for handling negative timestamps (before epoch), so this is also updated

File: Tests/RegoTests/BuiltinTests/TimeTests.swift (hunk 2 above)
Comment (DFrenkel) [reply Tests/RegoTests/BuiltinTests/TimeTests.swift:462]: ✅ Good catch - this was my miss.

**Comment by philipaconrad** [comment]
@DFrenkel I've assigned myself on this PR so that it won't go away on my notifications until it closes or merges.

File: Sources/Rego/Builtins/Time.swift (hunk 1 above)
Comment (philipaconrad) [reply Sources/Rego/Builtins/Time.swift]: I think the truncation behavior sounds reasonable, especially in context of `time.clock` / `time.date` / `time.weekday`, which are somewhat coarse/lossy ways of looking at a timestamp anyhow. Thanks for taking a second look at the timestamp handling!

**Review by philipaconrad — APPROVED** [review]
Thanks @DFrenkel for taking a second look at the timestamp handling, and for adding some extra edge-case tests. I think we're good to merge! 🙂
